src/main/scala/com/fulcrumgenomics/umi/UmiConsensusCaller.scala

```diff
@@ -0,0 +1,277 @@
+    def baseString = new String(bases)
+  }
+
+  /**
+    * Attempts to construct a String that can be used as a prefix for consensus read names based
+    * on the contents of the incoming SAMFileHeader.
+    */
+  def makePrefixFromSamHeader(header: SAMFileHeader): String = {
+    val ids = header.getReadGroups.map(rg => Option(rg.getLibrary).getOrElse(rg.getReadGroupId)).toList.sorted.distinct
+    // Read names have to fit into 255 bytes, and this is just the prefix
+    if (ids.map(_.length+1).sum <= 200) ids.mkString("|")
+    else Integer.toHexString(new Murmur3(1).hashUnencodedChars(ids.mkString("|")))
+  }
+
+  /**
+    * Constructs an output header with a single read group for the a BAM.
+    */
+  def outputHeader(in: SAMFileHeader, readGroupId: String, sortOrder: Option[SortOrder] = None): SAMFileHeader = {
+    val oldRgs = in.getReadGroups.toSeq
+    def collapse(f: SAMReadGroupRecord => String): String = oldRgs.map(f).filter(_ != null).toSet.toList match {
+      case Nil      => null
+      case x :: Nil => x
+      case xs       => xs.mkString(",")
+    }
+
+    val rg = new SAMReadGroupRecord(readGroupId)
+    rg.setDescription(s"Consensus reads generated from ${oldRgs.size} input read groups.")
+    rg.setLibrary         (collapse(_.getLibrary))
+    rg.setSample          (collapse(_.getSample))
+    rg.setPlatform        (collapse(_.getPlatform))
+    rg.setPlatformUnit    (collapse(_.getPlatformUnit))
+    rg.setSequencingCenter(collapse(_.getSequencingCenter))
+
+    val outHeader = new SAMFileHeader
+    outHeader.addReadGroup(rg)
+    outHeader.setSortOrder(sortOrder.getOrElse(SortOrder.unsorted))
+    outHeader.setGroupOrder(GroupOrder.query)
+    outHeader
+  }
+}
+
+
+/**
+  * A trait that can be mixed in by any consensus caller that works at the read level,
+  * mapping incoming SAMRecords into consensus SAMRecords.
+  *
+  * @tparam C Internally, the type of lightweight conesnsus read that is used prior to
+  *           rebuilding SAMRecords.
+  */
+trait UmiConsensusCaller[C <: ConsensusRead] {
+  import com.fulcrumgenomics.umi.UmiConsensusCaller.ReadType._
+
+  // vars to track how many reads meet various fates
+  private var _totalReads: Long = 0
+  private var _totalFiltered: Long = 0
+  private var _filteredForMinorityAlignment: Long = 0
+
+  /** Returns the total number of input reads examined by the consensus caller so far. */
+  def totalReads: Long = _totalReads
+
+  /**
+    * Returns the number of raw reads filtered out due to there being insufficient reads present
+    * to build the necessary set of consensus reads.
+    */
+  def readsFilteredInsufficientSupport: Long = this._totalFiltered - this._filteredForMinorityAlignment
+
+  /** Returns the number of raw reads filtered out because their alignment disagreed with the majority alignment of
+    * all raw reads for the same source molecule
+    */
+  def readsFilteredMinorityAlignment: Long = _filteredForMinorityAlignment
```

Comment:
It's whatever the callers send to the `reject` method.  For duplexes I've been lazy and not registered the fragments as rejects because the doc says they are ignored.  For the vanilla caller, I believe it's everything that is unused.